File: 16_String-II/M038. Count and Say.java
// Count and Say
// https://leetcode.com/problems/count-and-say/
// Problem Statement:
/*
The count-and-say sequence is a sequence of digit strings defined by the recursive formula:
countAndSay(1) = "1"
countAndSay(n) is the run-length encoding of countAndSay(n - 1).

Run-length encoding (RLE) is a string compression method that works by
replacing consecutive identical characters (repeated 2 or more times) with
the concatenation of the character and the number marking the count of the
characters (length of the run). For example, to compress the string "3322251"
we replace "33" with "23", replace "222" with "32", replace "5" with "15" and
replace "1" with "11". Thus the compressed string becomes "23321511".

Given a positive integer n, return the nth element of the count-and-say sequence.

Example:

Input: n = 4
Output: "1211"

Explanation:
    countAndSay(1) = "1"
    countAndSay(2) = RLE of "1" = "11"
    countAndSay(3) = RLE of "11" = "21"
    countAndSay(4) = RLE of "21" = "1211"

*/

class Solution {

    public String countAndSay(int n) {
        if (n == 1) return "1";
        String prev = countAndSay(n - 1);
        StringBuilder sb = new StringBuilder();
        int count = 1;
        char prevChar = prev.charAt(0);
        for (int i = 1; i < prev.length(); i++) {
            if (prev.charAt(i) == prevChar) {
                count++;
            } else {
                sb.append(count).append(prevChar);
                count = 1;
                prevChar = prev.charAt(i);
            }
        }
        sb.append(count).append(prevChar);
        return sb.toString();
    }

    public static void main(String[] args) {
        Solution solution = new Solution();
        System.out.println(solution.countAndSay(4));
    }
}

File: JavaScript/16_String-II/M038. Count and Say.js
// 38. Count and Say
// https://leetcode.com/problems/count-and-say/

// The count-and-say sequence is a sequence of digit strings defined by:
// countAndSay(1) = "1"
// countAndSay(n) is the run-length encoding of countAndSay(n - 1)

// Example:
// countAndSay(1) = "1"
// countAndSay(2) = "11" (one 1)
// countAndSay(3) = "21" (two 1s)
// countAndSay(4) = "1211" (one 2, one 1)

/**
 * Generate the nth term of the count-and-say sequence
 * @param {number} n - Term number (1-indexed)
 * @returns {string} - The nth term
 */
function countAndSay(n) {
    if (n === 1) return "1";

    const prev = countAndSay(n - 1);
    let result = "";
    let count = 1;
    let prevChar = prev[0];

    for (let i = 1; i < prev.length; i++) {
        if (prev[i] === prevChar) {
            count++;
        } else {
            result += count.toString() + prevChar;
            count = 1;
            prevChar = prev[i];
        }
    }

    // Append the last group
    result += count.toString() + prevChar;

    return result;
}

// Iterative version
function countAndSayIterative(n) {
    let result = "1";

    for (let i = 2; i <= n; i++) {
        let next = "";
        let count = 1;
        let prevChar = result[0];

        for (let j = 1; j < result.length; j++) {
            if (result[j] === prevChar) {
                count++;
            } else {
                next += count.toString() + prevChar;
                count = 1;
                prevChar = result[j];
            }
        }

        next += count.toString() + prevChar;
        result = next;
    }

    return result;
}

// Test cases
console.log(countAndSay(1)); // "1"
console.log(countAndSay(2)); // "11"
console.log(countAndSay(3)); // "21"
console.log(countAndSay(4)); // "1211"
console.log(countAndSay(5)); // "111221"
console.log(countAndSay(6)); // "312211"